akka-http-tests/src/test/scala/akka/http/scaladsl/server/HttpAppSpec.scala

```diff
@@ -0,0 +1,185 @@
+      }
+
+      verifyRequest(host, port)
+
+      minimal.binding().isSuccess should ===(true)
+      minimal.binding().get.localAddress.getPort should ===(port)
+      minimal.binding().get.localAddress.getHostName should ===(host)
+
+      minimal.shutdownPromise.success(Unit)
+      Await.ready(server, Duration(1, TimeUnit.SECONDS))
+      server.isCompleted should ===(true)
+    }
+
+    "let get notified" when {
+      class SneakSever extends HttpApp {
+
+        val shutdownPromise = Promise[Unit]()
+
+        val postBindingCalled = new AtomicBoolean(false)
+        val postBindingFailureCalled = new AtomicBoolean(false)
+        val postShutdownCalled = new AtomicBoolean(false)
+
+        override protected def route: Route = path("foo") {
+          complete("bar")
+        }
+
+        override protected def waitForShutdownSignal(system: ActorSystem)(implicit ec: ExecutionContext): Future[Unit] = {
+          shutdownPromise.future
+        }
+
+        override protected def postHttpBindingFailure(cause: Throwable): Unit = postBindingFailureCalled.set(true)
+
+        override protected def postHttpBinding(binding: ServerBinding): Unit = postBindingCalled.set(true)
+
+        override protected def postServerShutdown(attempt: Try[Unit], system: ActorSystem): Unit = postShutdownCalled.set(true)
+      }
+
+      "shutting down" in {
+        val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
+
+        val sneaky = new SneakSever()
+
+        val server = Future {
+          sneaky.startServer(host, port, ServerSettings(ConfigFactory.load))
+        }
+
+        sneaky.postShutdownCalled.get() should ===(false)
+
+        sneaky.shutdownPromise.success(Unit)
+        Await.ready(server, Duration(1, TimeUnit.SECONDS))
+        server.isCompleted should ===(true)
+        eventually {
+          sneaky.postShutdownCalled.get() should ===(true)
+        }
+      }
+
+      "after binding is successful" in {
+        val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
+
+        val sneaky = new SneakSever()
+
+        val server = Future {
+          sneaky.startServer(host, port, ServerSettings(ConfigFactory.load))
+        }
+
+        verifyRequest(host, port)
+
+        sneaky.postBindingCalled.get() should ===(true)
+
+        sneaky.shutdownPromise.success(Unit)
+        Await.ready(server, Duration(1, TimeUnit.SECONDS))
+        server.isCompleted should ===(true)
+      }
+
+      "after binding is unsuccessful" in {
+        val sneaky = new SneakSever()
+
+        sneaky.startServer("localhost", 1, ServerSettings(ConfigFactory.load))
+
+        eventually {
+          sneaky.postBindingFailureCalled.get() should ===(true)
+        }
+      }
+
+    }
+
+  }
+
+  private def verifyRequest(host: String, port: Int) = {
+
+    val request = HttpRequest(uri = s"http://$host:$port/foo")
+
+    implicit val mat = ActorMaterializer()
+
+    eventually {
```

Comment:
Using it to ensure the server is correctly up and running. Without it, I got some test errors due to races